apps/frontend/src/app/Components/LevelSelect.tsx

```diff
@@ -17,40 +17,76 @@ import CustomNumberInput, {
   CustomNumberInputButtonGroupWrapper,
 } from './CustomNumberInput'
 import DropdownButton from './DropdownMenu/DropdownButton'
+import type { ICharacterTalent } from '@genshin-optimizer/gi-good'
 
 export default function LevelSelect({
   level,
   ascension,
-  setBoth,
+  dispatch,
+  talent,
   useLow = false,
   disabled = false,
 }: {
   level: number
   ascension: AscensionKey
-  setBoth: (action: { level?: number; ascension?: AscensionKey }) => void
+  dispatch: (action: {
+    level?: number
+    ascension?: AscensionKey
+    talent?: ICharacterTalent
+  }) => void
+  talent?: ICharacterTalent
   useLow?: boolean
   disabled?: boolean
 }) {
   const { t } = useTranslation('ui')
   const ascensionMaxLevels = useLow ? ascensionMaxLevelLow : ascensionMaxLevel
+  const setLevelAndAscensionWithTalentClamp = useCallback(
+    (newLevelAndAscension: {
+      level?: number
+      ascension: AscensionKey
+      talent?: ICharacterTalent
+    }) => {
+      const dispatchSettings = newLevelAndAscension
+      // If changing ascension for a Character, clamp their talent levels
+      if (talent) {
```

Comment:
This should be in our db verification code too
https://github.com/frzyc/genshin-optimizer/blob/461cdc02f4b099906265f169e4089ef41e9d1951/apps/frontend/src/app/Database/DataManagers/CharacterData.ts#L50

Reply:
Added in https://github.com/frzyc/genshin-optimizer/pull/1454/commits/ab499328f1a78d8600d0a556ce1b30a7d1cfa74c